### ISSUE TITLE
Type could not be inferred despite explicit @return

### ISSUE THREAD

**Issue by wagenet** [castwide/solargraph#420]
<img width="439" alt="image" src="https://user-images.githubusercontent.com/9835/109877747-3da9b400-7c28-11eb-9b5e-60f58e982cd3.png">

<img width="525" alt="image" src="https://user-images.githubusercontent.com/9835/109877770-469a8580-7c28-11eb-96cc-17ef212ebf0c.png">

Rewriting the method as follows resolves the error:

```rb
  def http_token
    # @type [String, nil]
    return unless (header = request.headers["Authorization"])

    header.split.last
  end
```

However, it does not seem that this should be necessary. My understanding is that if the type cannot be inferred explicit typing should be sufficient. Is the complaint actually about not being able to infer the type of `request.headers["Authorization"]`? If so, is it just highlighting the wrong line?

**Comment by castwide** [comment]
It looks like the return type inference is incorrect because it handles the `&` nil safety operator incorrectly. A simple reproduction:

```ruby
# @return [String, nil]
def http_token Problem reported here: #http_token return type could not be inferred
  # @type [Hash{String => String}]
  headers = {}
  headers['Authorization']&.split(' ')&.last
end
```

There's another related problem where a parameterized hash (e.g., `[Hash{String => String}]`) does not properly handled the possibility of a nil result from `Hash#[]`. The return type from `headers['Authorization']` should be `[String, nil]`, but it's inferred to be `[String]`.

Fixes for both are on the roadmap.

**Comment by castwide** [comment]
There's a fix for the `&` nil safety operator in `master`. It'll be included in the next release.

**Comment by castwide** [comment]
Released in v0.44.1.
